### PR TITLE
feat(dynamic-links): API to directly resolve dynamic link

### LINKED ISSUE #3802: Help needed - resolve embedded link
I'm scanning QR code containing FB Dynamic Link inside from withing my app. I get something like this `https://myapp.page.link/S8rq8wTHFE1KYj8D8`.
The question is how do I resolve the underlying URL/link that was used when this link was created? 

When I paste this link in a browser it makes some redirects and eventually shows the URL I'm interested in. How do I do the same with FB SDK?

### PR DESCRIPTION
**Pull request by mikehardy**
### Description

The dynamic links module has no API to resolve links directly, though both underlying SDKs have it, and it's a common use case (imagine you just scan a QR with a dynamic link inside your app, then how do you get the link?)

There is no corresponding JS API so I just wrapped the "universal" handler which can take short or long links

### Related issues

Fixes #3802 with help from @alex-mironov

### Release Summary

feat(dynamic-links): API to directly resolve dynamic link

### Checklist

- I read the [Contributor Guide](/CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
  - [x] Yes
- My change supports the following platforms;
  - [x] `Android`
  - [x] `iOS`
- My change includes tests;
  - [x] `e2e` tests added or updated in `packages/\*\*/e2e`
  - [ ] `jest` tests added or updated in `packages/\*\*/__tests__`
- [x] I have updated TypeScript types that are affected by my change.
- This is a breaking change;
  - [ ] Yes
  - [x] No - **but there should be one for DynamicLink.minimumAppVersion to have consistent type across platforms**

### Test Plan

I have added thorough e2e testing, with the hardest part being the error conditions and associated documentation of failure modes, which I *think* I got correct everywhere

Runs on iOS simulator and Android emulator without fault now but it took a while to get there...

---

Think `react-native-firebase` is great? Please consider supporting the project with any of the below:

- 👉 Star this repo on GitHub ⭐️
- 👉 Follow [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter

### CODE REVIEW

**Comment by mikehardy** [comment]
This will save you time in review if you're curious about the APIs:

- Android: https://firebase.google.com/docs/reference/android/com/google/firebase/dynamiclinks/FirebaseDynamicLinks#getDynamicLink(android.net.Uri)

- iOS: https://firebase.google.com/docs/reference/ios/firebasedynamiclinks/api/reference/Classes/FIRDynamicLinks#-handleuniversallink:completion:

JS does not have the dynamic link module at all

**Comment by mikehardy** [comment]
@alex-mironov here's the patch for the dynamic link resolve PR in patch-package form (just strip the .txt extension I added for github file upload) of this PR, the error handling is really important and the snippets you posted worked in the happy path but I'd back out your FDL mini-module and use this instead of you had a choice

[@react-native-firebase+dynamic-links+7.1.4.patch.txt](https://github.com/invertase/react-native-firebase/files/4808711/%40react-native-firebase%2Bdynamic-links%2B7.1.4.patch.txt)


This is what I'll be using as I implement the same feature in my app

**Comment by mikehardy** [comment]
I can't get CI to cooperate. They both pass (and pass locally) but on github not both at the same time.

**Comment by mikehardy** [comment]
Okay E2E worked, finally - which makes me happy since actually testing all the failure cases, handling them and then proving they were handled was probably 80% of the work here. @Salakar / @Ehesp this is good to go I think?

File: packages/dynamic-links/android/src/main/java/io/invertase/firebase/dynamiclinks/ReactNativeFirebaseDynamicLinksModule.java [hunk 1]
```diff
@@ -144,6 +144,34 @@ public void getInitialLink(Promise promise) {
...
+    try {
+      FirebaseDynamicLinks.getInstance()
+        .getDynamicLink(Uri.parse(link))
+        .addOnCompleteListener(task -> {
+          if (task.isSuccessful()) {
+            PendingDynamicLinkData linkData = task.getResult();
+            // Careful: link == null if link invalid, isSuccessful is only false on processing error
+            if (linkData != null && linkData.getLink() != null && linkData.getLink().toString() != null) {
+              String linkUrl = linkData.getLink().toString();
+              // I can't believe they made this an int when versionCode is moving to long, but okay
```
Comment (Salakar) [inline packages/dynamic-links/android/src/main/java/io/invertase/firebase/dynamiclinks/ReactNativeFirebaseDynamicLinksModule.java:158]:
```suggestion
```

File: packages/dynamic-links/android/src/main/java/io/invertase/firebase/dynamiclinks/ReactNativeFirebaseDynamicLinksModule.java [hunk 2]
```diff
@@ -144,6 +144,33 @@ public void getInitialLink(Promise promise) {
...
 
+  @ReactMethod
+  public void resolveLink(String link, Promise promise) {
+    try {
+      FirebaseDynamicLinks.getInstance()
+        .getDynamicLink(Uri.parse(link))
+        .addOnCompleteListener(task -> {
+          if (task.isSuccessful()) {
+            PendingDynamicLinkData linkData = task.getResult();
+            // Careful: link == null if link invalid, isSuccessful is only false on processing error
```
Comment (Salakar) [inline packages/dynamic-links/android/src/main/java/io/invertase/firebase/dynamiclinks/ReactNativeFirebaseDynamicLinksModule.java:155]:
```suggestion
            // Note: link == null if link invalid, isSuccessful is only false on processing error
```

File: packages/dynamic-links/ios/RNFBDynamicLinks/RNFBDynamicLinksModule.m [hunk 3]
```diff
@@ -182,6 +182,34 @@ - (id)init {
...
+{
+  id completion = ^(FIRDynamicLink *_Nullable dynamicLink, NSError *_Nullable error) {
+    if (!error && dynamicLink && dynamicLink.url) {
+        resolve(@{
+            @"url": dynamicLink.url.absoluteString,
+            @"minimumAppVersion": dynamicLink.minimumAppVersion == nil ? [NSNull null] : dynamicLink.minimumAppVersion,
+        });
+    } else if (!error) {
+      [RNFBSharedUtils rejectPromiseWithUserInfo:reject userInfo:(NSMutableDictionary *) @{
+          @"code": @"resolve-link-not-found",
```
Comment (Salakar) [inline packages/dynamic-links/ios/RNFBDynamicLinks/RNFBDynamicLinksModule.m:198]:
```suggestion
          @"code": @"not-found",
```

File: packages/dynamic-links/lib/index.d.ts [hunk 4]
```diff
@@ -558,6 +562,28 @@ export namespace FirebaseDynamicLinksTypes {
...
+     *
+     * #### Example
+     *
+     * ```js
+     * const link = await firebase.dynamicLinks().resolveLink('https://reactnativefirebase.page.link/76adfasdf');
+     * console.log('Received link with URL: ' + link.url);
+     * ```
+     *
+     * Can throw error with message 'Invalid link parameter' if link parameter is null
+     * Can throw error with code 'resolve-link-not-found' if the link does not resolve
```
Comment (Salakar) [inline packages/dynamic-links/lib/index.d.ts:580]:
```suggestion
     * Can throw error with code 'not-found' if the link does not resolve
```

File: packages/dynamic-links/lib/index.js.flow [hunk 5]
```diff
@@ -534,6 +564,27 @@ export interface Module extends ReactNativeFirebaseModule {
...
+   * would get from {@link dynamic-links#getInitialLink}
+   *
+   * #### Example
+   *
+   * ```js
+   * const link = await firebase.dynamicLinks().resolveLink('https://invertase.io');
+   * ```
+   *
+   * Can throw error with message 'Invalid link parameter' if link parameter is null
+   * Can throw error with code 'resolve-link-not-found' if the link does not resolve
```
Comment (Salakar) [inline packages/dynamic-links/lib/index.js.flow:581]:
```suggestion
   * Can throw error with code 'not-found' if the link does not resolve
```

File: packages/dynamic-links/e2e/dynamicLinks.e2e.js [hunk 6]
```diff
@@ -80,6 +80,76 @@ describe('dynamicLinks()', () => {
...
+      // On Android it's a number and iOS a String, so parseInt is used to have a single test
+      parseInt(link.minimumAppVersion, 10).should.equal(123);
+    });
+
+    it('throws on links that do not exist', async () => {
+      try {
+        await firebase.dynamicLinks().resolveLink(baseParams.domainUriPrefix + '/not-a-valid-link');
+        return Promise.reject(new Error('Did not throw Error.'));
+      } catch (e) {
+        e.code.should.containEql('resolve-link-not-found');
```
Comment (Salakar) [inline packages/dynamic-links/e2e/dynamicLinks.e2e.js:124]:
```suggestion
        e.code.should.containEql('not-found');
```

File: packages/dynamic-links/android/src/main/java/io/invertase/firebase/dynamiclinks/ReactNativeFirebaseDynamicLinksModule.java [hunk 7]
```diff
@@ -144,6 +144,33 @@ public void getInitialLink(Promise promise) {
...
+        .addOnCompleteListener(task -> {
+          if (task.isSuccessful()) {
+            PendingDynamicLinkData linkData = task.getResult();
+            // Note: link == null if link invalid, isSuccessful is only false on processing error
+            if (linkData != null && linkData.getLink() != null && linkData.getLink().toString() != null) {
+              String linkUrl = linkData.getLink().toString();
+              int linkMinimumVersion = linkData.getMinimumAppVersion();
+              promise.resolve(dynamicLinkToWritableMap(linkUrl, linkMinimumVersion));
+            } else {
+              rejectPromiseWithCodeAndMessage(promise, "resolve-link-not-found", "Dynamic link not found");
```
Comment (Salakar) [inline packages/dynamic-links/android/src/main/java/io/invertase/firebase/dynamiclinks/ReactNativeFirebaseDynamicLinksModule.java:161]:
```suggestion
              rejectPromiseWithCodeAndMessage(promise, "not-found", "Dynamic link not found");
```

File: packages/dynamic-links/android/src/main/java/io/invertase/firebase/dynamiclinks/ReactNativeFirebaseDynamicLinksModule.java (hunk 1 above)
Comment (mikehardy) [reply packages/dynamic-links/android/src/main/java/io/invertase/firebase/dynamiclinks/ReactNativeFirebaseDynamicLinksModule.java]: Yeah this was definitely a dev note - now at least you know that you'll see this line of code again though, when they move to long

**Comment by mikehardy** [comment]
@alex-mironov updated patch after review, this looks the version that will merge for release, only change from previous is when link is not found error code is 'not-found' vs 'resolve-link-not-found'

[@react-native-firebase+dynamic-links+7.1.4.patch.txt](https://github.com/invertase/react-native-firebase/files/4814047/%40react-native-firebase%2Bdynamic-links%2B7.1.4.patch.txt)

Thanks again for the collaboration finding the APIs and providing the base native implementations, really sped up the process when I went to wrap them with JS here and add tests :fist_right: :fist_left:

**Comment by mikehardy** [comment]
@Salakar thanks for merging this!

**Comment by Salakar** [comment]
Thanks for the thanks @mikehardy :D
-- 





-

The information in this email and in any attachments is 
confidential and may be privileged. If you are not the intended recipient, 
please destroy this message, delete any copies held on your systems and 
notify the sender immediately. Unauthorised copying, use or disclosure of 
information contained herein is strictly prohibited. Any personal content 
in this email is that of the individual author and should not be 
interpreted as being endorsed by the company in whole or in part. Whilst 
care is taken, it is the responsibility of the recipient to ensure that 
this email is free from virus infection, and no responsibility is accepted 
by Invertase Limited for any disruption, loss or damage arising from its 
receipt or use.****

*** ***

P Please consider the environment before 
printing this email

**Comment by Salakar** [comment]
Test

**Pull request merged**